### DISCUSSION TITLE
Create Path from curvature

### DISCUSSION THREAD

**Discussion by kazimgorgulu** [gdsfactory/gdsfactory#2780] [Q&A]
Hi everyone,

I want to create custom curvature bends and therefore I need to find a way to get `Path` from a given curvature and arc-length.
I created a function for this purpose:

```python
import gdsfactory as gf
import numpy as np

def calc_path_from_curvature(s, K):
    """
    s: arc-length array
    K: curvature array
    """
    ds = np.gradient(s, edge_order=2)
    dtheta = K*ds
    theta = np.concatenate([[0], np.cumsum(dtheta)])
    theta = np.convolve(theta, np.ones(2), 'valid')/2
    
    dx = ds*np.cos(theta)
    dy = ds*np.sin(theta)

    xpoints = np.concatenate([[0.], np.cumsum(dx)])
    ypoints = np.concatenate([[0.], np.cumsum(dy)])

    points = np.column_stack([xpoints, ypoints])
    path = gf.path.Path()
    path.append(points)
    return path, theta[-1]*180/np.pi
```

When I try to test this function with existing path functions I sometimes get slightly different paths:

```python
P = gf.path.arc(radius=16, angle=90, npoints=600)
s = P.curvature()[0]
K = P.curvature()[1]
PP, end_angle = calc_path_from_curvature(s, K)

print(P.points[-1])
print(PP.points[-1])
print(end_angle)

Outputs:
>> [16. 16.]
>> [16.02096516 15.97900734]
>> 89.92487479132004
```
The difference decreases as I increase `npoints` which I think indicates decreasing numerical error during integration in my function.

arc-length (s) and curvature (K) of a path are calculated numerically in `Path` so I think it should be possible to reverse this process and get the exact path back again from these parameters. Where do you think I am doing wrong?

**Comment by sebastian-goeldi** [comment]
The problem is that your code would come with some assumptions which are not true in GDS/OASIS

- gdsfactory works on numpy floats (i.e. most likely np.float64)
- gds/oas can take arbitrary precision

Both are not true.

- klayout floatings are either 32 or 64 bit precision (likely 64) floats that are mapped to native python floats
- gds and oasis are working on integer grids with a database unit (likely 1nm)

The path is most likely already snapped to grid and therefore constrained to precisions of the grid when converting back to µm whic h is why you get nicer numbers when not using the numpy approximation.
